### PR TITLE
Generate: TF `compute_transition_scores`

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

This PR adds the TF `compute_transition_scores`, akin to PT's #21191.

What seemingly started off as a simple task, ended up being a complex task -- the TF side had many missing and/or incorrect secondary `generate` outputs 😬 This means that we need to beef up the TF side of `generate` tests, which is much shorter than its PT counterpart.

### CODE REVIEW

File: src/transformers/generation/tf_utils.py [hunk 1]
```diff
@@ -480,6 +495,126 @@ def adjust_logits_during_generation(
...
+                `(batch_size*num_return_sequences, sequence_length)`. Only required if a `num_beams>1` at
+                generate-time.
+            normalize_logits (`bool`, *optional*, defaults to `False`):
+                Whether to normalize the logits (which, for legacy reasons, may be unnormalized).
+
+        Return:
+            `tf.Tensor`: A `tf.Tensor` of shape `(batch_size*num_return_sequences, sequence_length)` containing
+                the transition scores (logits)
+
+        Examples:
```
Comment (gante) [inline src/transformers/generation/tf_utils.py:529]: It should be noted that this example is exactly the same as the PT's example, except for the TF-specific model classes and some rounding differences ;)

File: src/transformers/generation/tf_utils.py [hunk 2]
```diff
@@ -849,6 +984,7 @@ def generate(
                 length_penalty=generation_config.length_penalty,
                 early_stopping=generation_config.early_stopping,
                 logits_processor=logits_processor,
+                output_scores=generation_config.output_scores,
```
Comment (gante) [inline src/transformers/generation/tf_utils.py:987]: (this one was missing, which means that users of previous versions couldn't return the scores with beam search 🤦 )

File: src/transformers/generation/tf_utils.py [hunk 3]
```diff
@@ -1385,10 +1522,13 @@ def greedy_search_body_fn(generated, finished_sequences, cur_len, model_kwargs):
...
             next_token_logits = model_outputs.logits[:, -1]
 
+            # pre-process distribution
+            next_tokens_scores = logits_processor(generated, next_token_logits, cur_len)
+
             # Store scores, attentions and hidden_states when required
             if not use_xla and return_dict_in_generate:
                 if output_scores:
-                    scores.append(next_token_logits)
+                    scores.append(next_tokens_scores)
```
Comment (gante) [inline src/transformers/generation/tf_utils.py:1531]: We want `scores` to contain the post-`logits_processor` logits.

File: src/transformers/generation/tf_utils.py [hunk 4]
```diff
@@ -1982,6 +2119,10 @@ def unflatten_beam_dim(tensor, num_beams, batch_axis=0):
         )
         scores = tf.ones((batch_size, num_beams)) * -1.0e9
 
+        # per batch beam indices
+        running_beam_indices = tf.ones((batch_size, num_beams, max_length), dtype=tf.int32) * -1
+        beam_indices = tf.ones((batch_size, num_beams, max_length), dtype=tf.int32) * -1
```
Comment (gante) [inline src/transformers/generation/tf_utils.py:2124]: These variables will collect the selected beam indices, during beam search. This information is needed to compute the scores in `compute_transition_scores`

File: src/transformers/generation/utils.py [hunk 5]
```diff
@@ -3068,14 +3066,14 @@ def beam_sample(
                 next_token_logits, dim=-1
             )  # (batch_size * num_beams, vocab_size)
 
-            next_token_scores_processed = logits_processor(input_ids, next_token_scores)
-            next_token_scores = next_token_scores_processed + beam_scores[:, None].expand_as(next_token_scores)
+            next_token_scores = logits_processor(input_ids, next_token_scores)
             next_token_scores = logits_warper(input_ids, next_token_scores)
+            next_scores = next_token_scores + beam_scores[:, None].expand_as(next_token_scores)
```
Comment (gante) [inline src/transformers/generation/utils.py:3071]: @patrickvonplaten ⚠️ important detail here: `beam_sample` was applying `logits_warper` on the accumulated score, as opposed to the token score. 

Since the logits warpers are used to adapt the token-level log probs, I am assuming this is bug -- hence this change :)

File: tests/generation/test_tf_utils.py [hunk 6]
```diff
@@ -181,3 +183,152 @@ def test_validate_generation_inputs(self):
         with self.assertRaisesRegex(ValueError, "foo"):
             fake_model_kwargs = {"foo": "bar"}
             model.generate(input_ids, **fake_model_kwargs)
+
+    def test_transition_scores_greedy_search(self):
```
Comment (gante) [inline tests/generation/test_tf_utils.py:187]: For all tests added here: exact same tests and checks as in PT (including the numbers in the numerical checks) 🙏

File: tests/generation/test_utils.py [hunk 7]
```diff
@@ -2489,10 +2489,10 @@ def test_generate_encoder_outputs_attention_mask(self):
 
     def test_transition_scores_greedy_search(self):
         articles = ["Justin Timberlake", "Michael Phelps"]
-        tokenizer = GPT2Tokenizer.from_pretrained("hf-internal-testing/tiny-random-gpt2")
+        tokenizer = GPT2Tokenizer.from_pretrained("distilgpt2", padding_side="left")
```
Comment (gante) [inline tests/generation/test_utils.py:2492]: Changed this one, as the random model had tiny numerical differences that made it select a different token in TF 😠

File: tests/generation/test_utils.py [hunk 8]
```diff
@@ -2533,11 +2533,11 @@ def test_transition_scores_greedy_search_normalized(self):
...
         expected_scores = np.array(
             [
-                [-6.5532393, -6.5158753, -6.451863, -6.4527144, -6.459402],
-                [-6.5685124, -6.4277077, -6.282607, -6.399295, -6.340927],
+                [-2.538938, -2.2694316, -2.1580915, -1.572299, -2.6719835],
+                [-1.8826028, -2.2461371, -1.7556462, -2.9644494, -1.7996008],
             ]
         )
-        self.assertTrue(np.allclose(transition_scores.cpu().numpy(), expected_scores))
+        self.assertTrue(np.allclose(transition_scores.cpu().numpy(), expected_scores, atol=1e-4))
```
Comment (gante) [inline tests/generation/test_utils.py:2540]: `atol` needed for CPU<>GPU output mismatches (the hardcoded values are correct on GPU)

File: tests/generation/test_tf_utils.py (hunk 6 above)
Comment (sgugger) [reply tests/generation/test_tf_utils.py:187]: Nice!

**Review by sgugger — APPROVED** [review]
Thanks for working on this!

File: src/transformers/generation/utils.py [hunk 9]
```diff
@@ -301,9 +301,9 @@ class BeamSearchDecoderOnlyOutput(ModelOutput):
             of log probabilities of tokens conditioned on log softmax of previously generated tokens in this beam.
             Tuple of `torch.FloatTensor` with up to `max_new_tokens` elements (one element for each generated token),
             with each tensor of shape `(batch_size*num_beams*num_return_sequences, config.vocab_size)`.
-        beam_indices (`tuple(tuple(torch.LongTensor))`, *optional*, returned when `output_scores=True` is passed or when `config.output_scores=True`):
+        beam_indices (`torch.LongTensor`, *optional*, returned when `output_scores=True` is passed or when `config.output_scores=True`):
             Beam indices of generated token id at each generation step. `torch.LongTensor` of shape
-            `(batch_size*num_return_sequences, input_ids.shape[-1])`.
+            `(batch_size*num_return_sequences, sequence_length)`.
```
Comment (patrickvonplaten) [inline src/transformers/generation/utils.py:306]: Guess it's a bit difficult for the user to compute what `sequence_length` is, but `input_ids.shape[-1]` is also not great. I wonder whether it could make sense to add a small doc that states how `sequence_length` is computed depending on the different use cases? Given how popular `generate` is, might be worth it ?

File: src/transformers/generation/utils.py (hunk 5 above)
Comment (patrickvonplaten) [reply src/transformers/generation/utils.py:3071]: This was actually intentional, we never changed it simply because it's backward breaking and because it's not clear that `logits_warper` should be applied **before** adding the beam_scores. 

Note that `logits_warper` is not the same as `logits_processor`. Logits warper includes top_p, top_k etc... those filters are supposed to be applied directly **before** the softmax . Note that top_k = 50 means that softmax can only pick from 50 values from the logits tensor (see blog post here: https://huggingface.co/blog/how-to-generate#top-k-sampling).

=> So I'm not in favor of this change. It's an exotic case as beam sample is a semi-official generation method, but IMO it's better to apply the logits warper before the softmax because all its parameters (temp, top_p, top_k) are tailored to be applied directly before the softmax. Also it's breaking if we change it here

File: tests/generation/test_utils.py (hunk 7 above)
Comment (patrickvonplaten) [reply tests/generation/test_utils.py:2492]: Haha I feel you!

**Review by patrickvonplaten — COMMENTED** [review]
Super nice tests! I think we need to revert the logits warper change though for PT. How is it done for TF - are they currently in sync?

File: src/transformers/generation/utils.py (hunk 5 above)
Comment (patrickvonplaten) [reply src/transformers/generation/utils.py:3071]: This was previously already discussed here as a reference: https://github.com/huggingface/transformers/pull/5420#discussion_r449779867

File: src/transformers/generation/utils.py (hunk 9 above)
Comment (gante) [reply src/transformers/generation/utils.py:306]: Yes, a user asked what `sequence_length` means two days ago! I'll add a TODO to try to improve these docs

File: src/transformers/generation/utils.py [hunk 10]
```diff
@@ -3091,15 +3089,15 @@ def beam_sample(
                     )
 
             # reshape for beam search
-            vocab_size = next_token_scores.shape[-1]
-            next_token_scores = next_token_scores.view(batch_size, num_beams * vocab_size)
+            vocab_size = next_scores.shape[-1]
+            next_scores = next_scores.view(batch_size, num_beams * vocab_size)
```
Comment (patrickvonplaten) [inline src/transformers/generation/utils.py:3093]: This means the softmax is applied over multiple beam scores no?

**Comment by gante** [comment]
@patrickvonplaten reverted the change in the order of operations in beam sample, and added a comment to based on our offline conversation.

LMK if you're happy with the PR 🙏

**Review by patrickvonplaten — APPROVED** [review]
Thanks!
